text/0000-edition-2021.md

```diff
@@ -0,0 +1,238 @@
+    * Alternatively, if that scenario is deemed unlikely, the RFC can state that this sort of code will not be automatically ported between editions. It should then describe if it is possible to at least issue a warning that the code will no longer work or will change meaning in the new edition.
+
+## Tooling workflow
+
+The expected workflow for upgrading to a new edition is as follows:
+
+* **Prepare to upgrade:** Run `cargo fix --edition`: This will automatically prepare your code to be upgraded to next edition by applying any suggestions from the migration lints.
+    * For example, if your code is in Rust 2018, this will prepare you to move to the 2021 edition.
+    * Note that `cargo fix` does not actually move your code to the new edition. Instead, it produces code that works in both the old and the new edition.
+        * This allows people to upgrade and fix things in a "piecemeal" fashion. Because of this, however, the code produced by these suggestions is not always the most idiomatic, as it is not able to take advantage of features from the new edition.
+    * The expectation is that `cargo fix` should be able to fix the majority of crates out there, but it is not required that the tooling is able to handle every case. As long as code does not occur frequently in the wild, it is acceptable for the automated suggestions to be inapplicable to edge cases. The metrics section in this RFC includes some guidelines for how to measure this.
+* **Upgrade:** Edit your cargo.toml to include `edition = "2021"` instead of the older edition.
+    * The code should still build, but it may be necessary to make some fixes or other changes.
+* **Cleanup:** After upgrading, you should run `cargo fix` again. This second step will apply any remaining changes that are only possible in the new edition.
+    * In some cases, these lints may be cleaning up "non-idiomatic" code produced by a migration.
+* **Manual changes:** As the final step, resolve any remaining lints or errors manually (hopefully these will be quite limited). Idiom lints, in particular, may not have automated suggestions, though you can always add an `#![allow]` at the crate level to silence them if you are in a hurry.
+
+## Limits of edition changes
+
+Edition changes cannot make arbitrary changes. First and foremost, they must always respect the constraint that crates in different editions can interoperate; this implies that the changes must be "skin deep". In terms of the compiler implementation, what this means is that Edition can change the way that concrete Rust syntax is desugared into MIR (the compiler's internal representation for Rust code), but doesn't change the semantics of MIR itself (or trait matching, etc).
+
+Some examples of changes the editions can make:
+
+* Introducing a new keyword, as with `async-await`.
+    * This change comes with automated tooling that rewrites identifiers using that keyword to use `r#` form.
```

Comment:
I don't think this is great. Sure the code works but it's ugly. Yes, the proper fix requires developer interaction but it will require it anyway unless the developer is fine with ugly marker.

Having an option to replace it with some other specific identifier would be much nicer. I realize that it may need to be different in different contexts.

If this is too complicated (but maybe the code can be reused in a general-purpose refactoring tool?), then at least a message saying "Keyword foo was replaced in 5 scopes" would be helpful reminder.

Reply:
it would be nice if a preference list could be given, e.g. I've seen the pattern that variables which wanted to use internal keywords as a name, were given the name `{name}_` (e.g. with an appended underscore at the end) instead of `r#{name}` (which I've also seen, but less often).

Reply:
I meant having the option configurable. E.g. if you try to apply `cargo fix` it will fail "the code uses new keywords as identifiers" message and then the user would supply either:

* `--raw-identifiers`
* `--replace-ident async=asynchronous --replace-ident await=awaiting`

This would probably cover all major use cases. The only thing it can't do is replacing based on context (maybe the different replacement is better in different functions), but this seems like an acceptable limitation to me.

Reply:
That does sound like a nice option, but I would prefer not to add it as required by this RFC. We've added keywords in the past using the simpler system and it did seem adequate; one can also do a search-and-replace with `ripgrep` readily enough as well.

Reply:
My primary concern is people running cargo fix, they'll see it "works" and don't notice changed identifier. So I believe at least warning should be strongly encouraged.

Reply:
Hm. I think one of our key goals is that editions can be migrated to piecemeal, and ultimately in rarely read code, this is fine; asking users to add an allow for working code is not too typical in Rust. Regardless, I think this could be added without changes to this rfc, per usual lang team processes (as an extension to the existing lints around identifier naming); I am not inclined to specify a particular behavior here.

Reply:
I meant one-time warning emitted from `cargo fix`. (Something like "The identifier x, which is now a keyword, was converted to raw form 5 times, you may wish to rename it to something else.")

Reply:
I don't want to prescribe such specifics in this RFC, but I think it may make sense for migrations to have the ability to give some kind of report - I know the closure changes for example could tell you how much the closures in your code changed size, etc. which might be difficult otherwise. I'd say this feels like something you could file an issue for on rust-lang/rust, though, just like any other diagnostic improvement, and we can discuss there.